pkg/apis/softwarecomposition/network_types.go

```diff
@@ -0,0 +1,65 @@
+package softwarecomposition
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Protocol string
+type CommunicationType string
+
+const (
+	ProtocolTCP  Protocol = "TCP"
+	ProtocolUDP  Protocol = "UDP"
+	ProtocolSCTP Protocol = "SCTP"
+
+	CommunicationTypeIngress CommunicationType = "internal"
+	CommunicationTypeEgress  CommunicationType = "external"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NetworkNeighborsList is a list of NetworkNeighbors.
+type NetworkNeighborsList struct {
+	metav1.TypeMeta
+	metav1.ListMeta
+
+	Items []NetworkNeighbors
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NetworkNeighbors represents a list of network communications for a specific workload.
+type NetworkNeighbors struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec NetworkNeighborsSpec
+}
+
+type NetworkNeighborsSpec struct {
+	metav1.LabelSelector // The labels which are inside spec.selector in the parent workload.
+	Ingress              []NetworkEntry
+	Egress               []NetworkEntry
+}
+
+// IngressEntry represents a single network communication.
+type NetworkEntry struct {
+	Identifier        string
+	Type              CommunicationType
+	DNS               string
+	Ports             []NetworkPort
+	PodSelector       *metav1.LabelSelector
+	NamespaceSelector *metav1.LabelSelector
+	IPAddress         string
+}
+
+type NetworkPort struct {
+	// Name is an artificial identifier of the network port. We use it for merging keys with Strategic Merge Patch.
+	// Format is `{protocol}-{port}`.
+	//
+	// Example: tcp-6881
+	Name     string // protocol-port
+	Protocol Protocol
+	Port     *int32
```

Comment:
question: why pointer to `int32`?

If you referenced what they do in K8s, the use a pointer to communicate that the value is optional. Is `Port` an optional value here?

Reply:
this differentiates between an empty port or a port set to zero

Reply:
Makes sense. Just to re-iterate our conversation in Slack, if we have it as a pointer, as you originally suggested, we can give a better error message when a client tries to create a NetworkNeighbor without a port value specified. I think that’s the way to do it.

It’s your call on whether to have `Port` as a pointer.